target/bin/addmailuser

```diff
@@ -73,3 +88,9 @@ then
     sleep 1
   done
 fi
+
+# Only necessary when user encryption keys are enabled
+if "${CRYPTOKEY_GENERATE}"
+then
+  doveadm -o plugin/mail_crypt_private_password="${PASSWD}" mailbox cryptokey generate -u "${FULL_EMAIL}" -U
```

Comment:
This line is failing on my test setup after running `./setup.sh email add -g testuser10@domain.tld newpassword` with:
`usage: doveadm [-Dv] [-f <formatter>] mailbox <command> [<args>] .....`
Manually running this command inside the docker container is working. Any idea whats different?

Reply:
Hmm, I get the feeling it's because the add command runs with 
```
        add      ) shift 2 ; _docker_image addmailuser "${@:+$@}" ;;

```
Instead of _docker_container (your existing container, dovecot version, etc).

@marhmm, are you using an unmodified version of the docker-mailserver? Or did you upgrade your dovecot version?

Reply:
Thanks for this hint, setting the `USE_CONTAINER=true` inside the setup.sh script fixes this behaviour!
I created a test docker image with a different name based on your branch and used this image name in the compose file.

Reply:
@marhmm Ah, yes, that makes sense. Running `./setup.sh -c {container_name}` would be a way to solve that without manually changing USER_CONTAINER.

@casperklein how do you feel about moving the addmailuser to default to _docker_container? I do want to acknowledge that it wouldn't have solved marhmm's issue because they used a non-default container name, but I do think it's a good thing to reify. I believe you and I had spoken about why we use _docker_image vs _container in a previous discussion. To me, using _docker_container is great if the user has modified their container tag (like using a newer docker version like I am).

Reply:
Okay, after cleaning up all my containers and config files and trying to understand what's happening in the setup.sh file I still think there is something missing:
Without the USE_CONTAINER=true setting, the setup.sh file will create a new (temporary) container using the image from the docker-compose file:
`${CRI} run --rm \
      -v "${CONFIG_PATH}:/tmp/docker-mailserver${USE_SELINUX}" \
      "${USE_TTY}" "${IMAGE_NAME}" "${@:+$@}"`

In contrast to the docker-compose file we don't mount the 10-custom.conf file here. In the result, dovecot does not have the cryptokey command included, which is needed by the doveadm command above.
Manually adding the bind mount option in the setup.sh file like in the docker-compose file fixes this.
I suggest to somehow configure dovecot to use the 10-custom.conf file directly from the "normal" config folder instead of bind mount it into the dovecot folder.
What do you think? Did I miss something?

Reply:
100% in agreement with your investigation and result. That's why I think we should move to use _docker_container by default for add and update. Let's see what some of the other devs think.

Reply:
I don't think we can use _docker_container for add. On the very first start of the container (with no existing config) dovecot does not start yet and we are stuck at `echo "Waiting for dovecot to create /var/mail/${DOMAIN}/${USER}..."` on line 87 in addmailuser

Reply:
I see two tests failing with the add script on _docker_container:

```
✗ checking setup.sh: setup.sh setquota
                                          (from function `assert_success' in file test/test_helper/bats-assert/src/assert.bash, line 114,
         in test file test/tests.bats, line 97)
                                                    `assert_success' failed
                                                                              
                                                                                 -- command failed --
                                                                                                        status : 1
                                                                                                                     output (15 lines):
        grep: /tmp/docker-mailserver/postfix-accounts.cf: No such file or directory
     Usage: setquota <user@domain> [<quota>]
     Error :: user quota_user@example.com does not exist
     Aborting.
     
          --- UNCHECKED ERROR
                                    - script    = setup.sh
                                                                 - function  = _docker_image / ${CRI} run --rm -v "${CONFIG_PATH}:/tmp/docker-mailserver${USE_SELINUX}" "${USE_TTY}" "${IMAGE_NAME}" "${@:+$@}"
                                                                                  - line      = 230
                                                                                                          - exit code = 1
                                                                                                                              
                                                                                                                                   Make sure you use a version of this script that matches
                                                           the version / tag of docker-mailserver. Please read the
                                                                                                                       'Get the tools' section in the README on GitHub careful-
                                                ly and use ./setup.sh help and read the VERSION section.
                                                                                                           --
                                                                                                                
 ✗ checking setup.sh: setup.sh delquota
                                          (from function `assert_success' in file test/test_helper/bats-assert/src/assert.bash, line 114,
         in test file test/tests.bats, line 125)
                                                     `assert_success' failed
                                                                               
                                                                                  -- command failed --
                                                                                                         status : 1
                                                                                                                      output (15 lines):
         grep: /tmp/docker-mailserver/postfix-accounts.cf: No such file or directory
     Usage: setquota <user@domain> [<quota>]
     Error :: user quota_user@example.com does not exist
     Aborting.
     
          --- UNCHECKED ERROR
                                    - script    = setup.sh
                                                                 - function  = _docker_image / ${CRI} run --rm -v "${CONFIG_PATH}:/tmp/docker-mailserver${USE_SELINUX}" "${USE_TTY}" "${IMAGE_NAME}" "${@:+$@}"
                                                                                  - line      = 230
                                                                                                          - exit code = 1
                                                                                                                              
                                                                                                                                   Make sure you use a version of this script that matches
                                                           the version / tag of docker-mailserver. Please read the
                                                                                                                       'Get the tools' section in the README on GitHub careful-
                                                ly and use ./setup.sh help and read the VERSION section.
                                                                                                           --
                                                                                                                

                                                                                                                2 tests, 2 failures
```

Reply:
Ok, yeah, this is getting a little complicated. There are a few tests that do `run ./setup.sh -p ./test/quota/config email add quota_user@example.com test_password` which means that ./test/quota/config won't be mounted into the container. I'm not entirely sure why `-p ./test/quota/config` is used when we can just use the existing config folder in the container.

I've removed `-p ./test/quota/config` from the tests and am using test/duplicate_configs/mail/ directly now and the tests work again.

Reply:
@marhmm, can you clarify the following for me with an example?

> On the very first start of the container (with no existing config) dovecot does not start yet and we are stuck at echo "Waiting for dovecot to create /var/mail/${DOMAIN}/${USER}..." on line 87 in addmailuser

What do I run to reproduce? Thanks!

Reply:
I tested it on commit 2bea9f44f81eeebb80154a8de7fa2e6b9a0cdbc5 of your issue branch (https://github.com/NorseGaud/docker-mailserver/commits/issues/2058). Built this image with the dockerfile and changed image, hostname and domainname in docker-compose.yml. Also added the two additional bind mounts (10-custom.conf and auth-passwdfile.inc).
After starting with docker-compose up I get the following:
```
--> sudo docker-compose up                                       
Starting mailserver ... done
Attaching to mailserver
mailserver    | [ TASKLOG ]  Welcome to docker-mailserver 10.0.0
mailserver    | [ TASKLOG ]  Initializing setup
mailserver    | [ TASKLOG ]  Checking configuration
mailserver    | [ TASKLOG ]  Configuring mail server
mailserver    | [ WARNING ]  Using self-generated dhparams is considered as insecure.
mailserver    | [ WARNING ]  Unless you known what you are doing, please remove /etc/dovecot/dh.pem.
mailserver    | [  FATAL  ]  Unless using LDAP, you need at least 1 email account to start Dovecot.
mailserver    | [  FATAL  ]  Please fix your configuration. Exiting...
```
This is fine for the first clean start I guess. So in a different terminal (without shutting down the first one) I run:
```
--> sudo ./setup.sh email add -g testuser1@domain.de newpass
Waiting for dovecot to create /var/mail/domain.de/testuser1...
Waiting for dovecot to create /var/mail/domain.de/testuser1...
Waiting for dovecot to create /var/mail/domain.de/testuser1...
^C
```
It is stuck here. Restarting the container will start the container completely, but the encryption step is missing then.

Reply:
That's caused by `while [[ ! -d "/var/mail/${DOMAIN}/${USER}" ]]` which means something is failing to create that directory.

Reply:
So this definitely won't work for the first user being created after running docker-compose up:

```
docker-compose up --build
Building mailserver
[+] Building 0.4s (43/43) FINISHED                                                                                                  
 => [internal] load build definition from Dockerfile                                                                           0.0s
 => => transferring dockerfile: 37B                                                                                            0.0s
 => [internal] load .dockerignore                                                                                              0.0s
 => => transferring context: 34B                                                                                               0.0s
 => [internal] load metadata for docker.io/library/debian:buster-slim                                                          0.0s
 => [ 1/39] FROM docker.io/library/debian:buster-slim                                                                          0.0s
 => [internal] load build context                                                                                              0.0s
 => => transferring context: 4.55kB                                                                                            0.0s
 => CACHED [ 2/39] RUN   echo 'deb http://deb.debian.org/debian buster-backports main' > /etc/apt/sources.list.d/buster-backp  0.0s
 => CACHED [ 3/39] RUN   echo '0 */6 * * * clamav /usr/bin/freshclam --quiet' >/etc/cron.d/clamav-freshclam &&   chmod 644 /e  0.0s
 => CACHED [ 4/39] COPY target/dovecot/auth-passwdfile.inc target/dovecot/??-*.conf /etc/dovecot/conf.d/                       0.0s
 => CACHED [ 5/39] COPY target/dovecot/sieve/ /etc/dovecot/sieve/                                                              0.0s
 => CACHED [ 6/39] COPY target/dovecot/dovecot-purge.cron /etc/cron.d/dovecot-purge.disabled                                   0.0s
 => CACHED [ 7/39] RUN chmod 0 /etc/cron.d/dovecot-purge.disabled                                                              0.0s
 => CACHED [ 8/39] WORKDIR /usr/share/dovecot                                                                                  0.0s
 => CACHED [ 9/39] RUN   sed -i -e 's/include_try \/usr\/share\/dovecot\/protocols\.d/include_try \/etc\/dovecot\/protocols\.  0.0s
 => CACHED [10/39] COPY target/dovecot/dovecot-ldap.conf.ext /etc/dovecot                                                      0.0s
 => CACHED [11/39] COPY   target/postfix/ldap-users.cf   target/postfix/ldap-groups.cf   target/postfix/ldap-aliases.cf   tar  0.0s
 => CACHED [12/39] RUN   sed -i -r 's/^(CRON)=0/\1=1/g' /etc/default/spamassassin &&   sed -i -r 's/^\$INIT restart/superviso  0.0s
 => CACHED [13/39] COPY   ./target/bin/*   ./target/scripts/*.sh   ./target/scripts/startup/*.sh   ./target/docker-configomat  0.0s
 => CACHED [14/39] COPY   ./VERSION /                                                                                          0.0s
 => CACHED [15/39] RUN   chmod +x /usr/local/bin/* &&   rm -rf /usr/share/locale/* &&   rm -rf /usr/share/man/* &&   rm -rf /  0.0s
 => CACHED [16/39] COPY target/postsrsd/postsrsd /etc/default/postsrsd                                                         0.0s
 => CACHED [17/39] COPY target/postgrey/postgrey /etc/default/postgrey                                                         0.0s
 => CACHED [18/39] COPY target/postgrey/postgrey.init /etc/init.d/postgrey                                                     0.0s
 => CACHED [19/39] RUN   chmod 755 /etc/init.d/postgrey &&   mkdir /var/run/postgrey &&   chown postgrey:postgrey /var/run/po  0.0s
 => CACHED [20/39] COPY target/amavis/conf.d/* /etc/amavis/conf.d/                                                             0.0s
 => CACHED [21/39] RUN   sed -i -r 's/#(@|   \\%)bypass/\1bypass/g' /etc/amavis/conf.d/15-content_filter_mode &&   adduser cl  0.0s
 => CACHED [22/39] RUN   for _ in {1..10}; do su - amavis -c "razor-admin -create" ; sleep 3 ;   if su - amavis -c "razor-adm  0.0s
 => CACHED [23/39] COPY target/fail2ban/jail.local /etc/fail2ban/jail.local                                                    0.0s
 => CACHED [24/39] RUN   ln -s /var/log/mail/mail.log /var/log/mail.log &&   rm /etc/fail2ban/jail.d/defaults-debian.conf &&   0.0s
 => CACHED [25/39] COPY target/opendkim/opendkim.conf /etc/opendkim.conf                                                       0.0s
 => CACHED [26/39] COPY target/opendkim/default-opendkim /etc/default/opendkim                                                 0.0s
 => CACHED [27/39] COPY target/opendmarc/opendmarc.conf /etc/opendmarc.conf                                                    0.0s
 => CACHED [28/39] COPY target/opendmarc/default-opendmarc /etc/default/opendmarc                                              0.0s
 => CACHED [29/39] COPY target/opendmarc/ignore.hosts /etc/opendmarc/ignore.hosts                                              0.0s
 => CACHED [30/39] RUN   update-alternatives --set iptables /usr/sbin/iptables-legacy &&   update-alternatives --set ip6table  0.0s
 => CACHED [31/39] COPY target/fetchmail/fetchmailrc /etc/fetchmailrc_general                                                  0.0s
 => CACHED [32/39] COPY target/postfix/main.cf target/postfix/master.cf /etc/postfix/                                          0.0s
 => CACHED [33/39] COPY target/shared/ffdhe4096.pem /etc/postfix/shared/ffdhe4096.pem                                          0.0s
 => CACHED [34/39] COPY   target/postfix/header_checks.pcre   target/postfix/sender_header_filter.pcre   target/postfix/sende  0.0s
 => CACHED [35/39] RUN   : >/etc/aliases &&   sed -i 's/START_DAEMON=no/START_DAEMON=yes/g' /etc/default/fetchmail &&   mkdir  0.0s
 => CACHED [36/39] RUN   sed -i -r "/^#?compress/c\compress\ncopytruncate" /etc/logrotate.conf &&   mkdir -p /var/log/mail &&  0.0s
 => CACHED [37/39] COPY target/supervisor/supervisord.conf /etc/supervisor/supervisord.conf                                    0.0s
 => CACHED [38/39] COPY target/supervisor/conf.d/* /etc/supervisor/conf.d/                                                     0.0s
 => exporting to image                                                                                                         0.1s
 => => exporting layers                                                                                                        0.0s
 => => writing image sha256:eb53b409ddb1d11ebcc11e6e9b2527ef42a57735214f06d536581dc8f05f2a3b                                   0.0s
 => => naming to docker.io/library/docker-mailserver_mailserver                                                                0.0s

Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them
mailserver is up-to-date
Attaching to mailserver
mailserver    | [ TASKLOG ]  Welcome to docker-mailserver 10.0.0
mailserver    | [ TASKLOG ]  Initializing setup
mailserver    | [ TASKLOG ]  Checking configuration
mailserver    | [ TASKLOG ]  Configuring mail server
mailserver    | sievec(root)<42><>: Debug: Loading modules from directory: /usr/lib/dovecot/modules
mailserver    | sievec(root)<42><>: Debug: Module loaded: /usr/lib/dovecot/modules/lib10_mail_crypt_plugin.so
mailserver    | sievec(root)<42><>: Debug: Module loaded: /usr/lib/dovecot/modules/lib10_quota_plugin.so
mailserver    | sievec: Debug: Effective uid=0, gid=0, home=/root
mailserver    | sievec: Debug: Quota root: name=User quota backend=count args=
mailserver    | sievec: Debug: Quota rule: root=User quota mailbox=* bytes=134217728 messages=0
mailserver    | sievec: Debug: Quota rule: root=User quota mailbox=Trash bytes=+52428800 messages=0
mailserver    | sievec: Debug: Quota warning: bytes=127506841 (95%) messages=0 reverse=no command=quota-warning 95 root 
mailserver    | sievec: Debug: Quota warning: bytes=107374182 (80%) messages=0 reverse=no command=quota-warning 80 root 
mailserver    | sievec: Debug: Quota warning: bytes=134217728 (100%) messages=0 reverse=yes command=quota-warning below root 
mailserver    | sievec: Debug: Quota grace: root=User quota bytes=13421772 (10%)
mailserver    | [  FATAL  ]  Unless using LDAP, you need at least 1 email account to start Dovecot.
mailserver    | [  FATAL  ]  Please fix your configuration. Exiting...
```

```
./setup.sh email add -g test@test.com test123
doveadm(test@test.com)<111><>: Error: userdb lookup: connect(/var/run/dovecot/auth-userdb) failed: No such file or directory
doveadm(test@test.com): Error: User lookup failed: Internal error occurred. Refer to server log for more information.

--- UNCHECKED ERROR
  - script    = setup.sh
  - function  = _docker_container / ${CRI} exec "${USE_TTY}" "${CONTAINER_NAME}" "${@:+$@}"
  - line      = 250
  - exit code = 75

Make sure you use a version of this script that matches
the version / tag of docker-mailserver. Please read the
'Get the tools' section in the README on GitHub careful-
ly and use ./setup.sh help and read the VERSION section.
```

However, the second user it does work:

```
./setup.sh email add -g test2@test.com test123
Waiting for dovecot to create /var/mail/test.com/test2...
   Folder    Public ID                                                                                  
✓  <userkey> 412989ffc6338bdeef7826a27fdb5f7b77a4b28483a10a351237243650e8e1bb                           

```

Reply:
> I tested it on commit [2bea9f4](https://github.com/docker-mailserver/docker-mailserver/commit/2bea9f44f81eeebb80154a8de7fa2e6b9a0cdbc5) of your issue branch (https://github.com/NorseGaud/docker-mailserver/commits/issues/2058). Built this image with the dockerfile and changed image, hostname and domainname in docker-compose.yml. Also added the two additional bind mounts (10-custom.conf and auth-passwdfile.inc).
> After starting with docker-compose up I get the following:
> 
> ```
> --> sudo docker-compose up                                       
> Starting mailserver ... done
> Attaching to mailserver
> mailserver    | [ TASKLOG ]  Welcome to docker-mailserver 10.0.0
> mailserver    | [ TASKLOG ]  Initializing setup
> mailserver    | [ TASKLOG ]  Checking configuration
> mailserver    | [ TASKLOG ]  Configuring mail server
> mailserver    | [ WARNING ]  Using self-generated dhparams is considered as insecure.
> mailserver    | [ WARNING ]  Unless you known what you are doing, please remove /etc/dovecot/dh.pem.
> mailserver    | [  FATAL  ]  Unless using LDAP, you need at least 1 email account to start Dovecot.
> mailserver    | [  FATAL  ]  Please fix your configuration. Exiting...
> ```
> 
> This is fine for the first clean start I guess. So in a different terminal (without shutting down the first one) I run:
> 
> ```
> --> sudo ./setup.sh email add -g testuser1@domain.de newpass
> Waiting for dovecot to create /var/mail/domain.de/testuser1...
> Waiting for dovecot to create /var/mail/domain.de/testuser1...
> Waiting for dovecot to create /var/mail/domain.de/testuser1...
> ^C
> ```
> 
> It is stuck here. Restarting the container will start the container completely, but the encryption step is missing then.

I finally reproduced it! Has to do with check-for-changes.sh not starting. https://github.com/docker-mailserver/docker-mailserver/pull/2080/commits/6ebf6647346ecc73f014b01b4a9327d33fef6dfd fixes it

Reply:
> @casperklein how do you feel about moving the addmailuser to default to _docker_container? I do want to acknowledge that it wouldn't have solved marhmm's issue because they used a non-default container name, but I do think it's a good thing to reify. I believe you and I had spoken about why we use _docker_image vs _container in a previous discussion. To me, using _docker_container is great if the user has modified their container tag (like using a newer docker version like I am).

See my summary here: https://github.com/docker-mailserver/docker-mailserver/pull/1874#issuecomment-809781531 --> If you want to force using a running container, you can always use ``-c``.

I saw, you've added comments for the dovecot community repo in the Dockerfile 👍. But be warned, that the comment was [intentionally removed in the past](https://github.com/docker-mailserver/docker-mailserver/pull/1626).

Reply:
Thanks! I think that it's way less code complexity to switch to defaulting to _docker_container with a fall back to _docker_image if it's not running then. This covers the "So you can for example create new mailboxes, before having a fully configured/running docker-mailserver yet." but also allows for custom changes in the container to be used should the user have made any.